src/target/riscv/riscv-013.c

```diff
@@ -3396,311 +3396,532 @@ static int write_memory_abstract(struct target *target, target_addr_t address,
+ * Prior to calling it, the program buffer should contain the appropriate
+ * program.
+ * This function sets DM_ABSTRACTAUTO_AUTOEXECDATA to trigger second stage of the
+ * pipeline (s1 -> dm_data[0:1]) whenever dm_data is read.
  */
-static int read_memory_progbuf_inner(struct target *target, target_addr_t address,
-		uint32_t size, uint32_t count, uint8_t *buffer, uint32_t increment)
+static int read_memory_progbuf_inner_startup(struct target *target,
+		target_addr_t address, uint32_t increment, uint32_t index)
 {
 	RISCV013_INFO(info);
+	/* s0 holds the next address to read from
+	 * s1 holds the next data value read
+	 * a0 is a counter in case increment is 0
+	 */
+	if (register_write_direct(target, GDB_REGNO_S0, address + index * increment)
+			!= ERROR_OK)
+		return ERROR_FAIL;
 
-	int result = ERROR_OK;
+	if (/*is_repeated_read*/ increment == 0 &&
+			register_write_direct(target, GDB_REGNO_A0, index) != ERROR_OK)
+		return ERROR_FAIL;
 
-	/* Write address to S0. */
-	result = register_write_direct(target, GDB_REGNO_S0, address);
-	if (result != ERROR_OK)
-		return result;
+	const uint32_t startup_command =
+		read_memory_progbuf_inner_startup_command(target);
+	if (execute_abstract_command(target, startup_command) != ERROR_OK)
+		return ERROR_FAIL;
 
-	if (increment == 0 &&
-			register_write_direct(target, GDB_REGNO_A0, 0) != ERROR_OK)
+	/* First read has just triggered. Result is in s1.
+	 * dm_data registers contain the previos value of s1 (garbage).
+	 */
+	if (dmi_write(target, DM_ABSTRACTAUTO,
+				set_field(0, DM_ABSTRACTAUTO_AUTOEXECDATA, 1)) != ERROR_OK)
 		return ERROR_FAIL;
 
-	uint32_t command = access_register_command(target, GDB_REGNO_S1,
-			riscv_xlen(target),
-			AC_ACCESS_REGISTER_TRANSFER | AC_ACCESS_REGISTER_POSTEXEC);
-	if (execute_abstract_command(target, command) != ERROR_OK)
+	/* Read garbage from dm_data0, which triggers another execution of the
+	 * program. Now dm_data contains the first good result (from s1),
+	 * and s1 the next memory value.
+	 */
+	if (dmi_read_exec(target, NULL, DM_DATA0) != ERROR_OK) {
+		dmi_write(target, DM_ABSTRACTAUTO, 0);
 		return ERROR_FAIL;
+	}
 
-	/* First read has just triggered. Result is in s1. */
-	if (count == 1) {
-		uint64_t value;
-		if (register_read_direct(target, &value, GDB_REGNO_S1) != ERROR_OK)
-			return ERROR_FAIL;
-		buf_set_u64(buffer, 0, 8 * size, value);
-		log_memory_access64(address, value, size, true);
-		return ERROR_OK;
+	uint32_t abstractcs;
+
+	if (wait_for_idle(target, &abstractcs))
+		return ERROR_FAIL;
+
+	info->cmderr = get_field(abstractcs, DM_ABSTRACTCS_CMDERR);
+	switch (info->cmderr) {
+	case CMDERR_NONE:
+		break;
+		LOG_TARGET_ERROR(target, "Unexpected busy error. This is probably a hardware bug.");
+		/* fall through */
+	default:
+		LOG_TARGET_DEBUG(target, "error when reading memory, cmderr=0x%" PRIx32,
+				info->cmderr);
+		riscv013_clear_abstract_error(target);
+		dmi_write(target, DM_ABSTRACTAUTO, 0);
+		return ERROR_FAIL;
```

Comment:
This looks as a typo to me - probably like a missing line `case CMDERR_BUSY:` 

```suggestion
	case CMDERR_NONE:
		break;
	case CMDERR_BUSY:
		LOG_TARGET_ERROR(target, "Unexpected busy error. This is probably a hardware bug.");
		/* fall through */
	default:
		LOG_TARGET_DEBUG(target, "error when reading memory, cmderr=0x%" PRIx32,
				info->cmderr);
		riscv013_clear_abstract_error(target);
		dmi_write(target, DM_ABSTRACTAUTO, 0);
		return ERROR_FAIL;
```

Reply:
Addressed. Thanks for the catch!